hacspec/src/poly.rs

```diff
@@ -267,38 +267,118 @@ pub(crate) fn extended_euclid_invert<T: TRestrictions<T>>(x: T, n: T, signed: bo
     t
 }
 
-/// Extended euclidean algorithm to compute the inverse of x in yℤ[x]
-#[inline]
-pub fn extended_euclid<T: TRestrictions<T>>(x: &[T], y: &[T], n: T) -> Result<Vec<T>, &'static str> {
-    let (x, y) = normalize(x, y);
-    println!("euclid: {:?} / {:?}", x, y);
-
-    let mut new_t = vec![T::default(); x.len()];
-    new_t[0] = T::from_literal(1);
+// Pad to degree d and reverse coefficients
+fn reverse<T: TRestrictions<T>>(x: &[T], d: usize) -> Vec<T> {
+    let mut out = vec![T::default(); d+1];
+    for i in 0..std::cmp::min(x.len(), d+1) {
+        out[d-i] = x[i]; //XXX: this could probably be written more nicely with iterators?
+    }
+    out 
+}
 
-    let mut new_r = x.clone();
+/// Subtract quotient (bn/x^bd) from (an/x^ad)
+fn quot_sub<T: TRestrictions<T>>(an: &[T], ad: usize, bn: &[T], bd: usize, n: T) -> (Vec<T>, usize){
+    let cd = std::cmp::max(ad, bd);
+    let x = monomial(T::from_literal(1),1);
+    let mut a = an.to_vec();
+    let mut b = bn.to_vec();
+    for _ in 0..cd-ad {
+        a = poly_mul(&a, &x, n);
+    }
+    for _ in 0..cd-bd {
+        b = poly_mul(&b, &x, n);
+    }
+    (poly_sub(&a, &b, n), cd)
+}
 
-    let mut t = vec![T::default(); x.len()];
-    let mut r = y.clone();
+/// Divide a by x assuming a is a multiple of x (shift right by one)
+fn poly_divx<T: TRestrictions<T>>(v: &[T]) -> Vec<T> {
+    let mut out = vec![T::default(); v.len()-1];
+    for i in 0..out.len() {
+        out[i] = v[i+1]; //XXX: this could probably be written more nicely with iterators?
```

Comment:
Something like this should work, or you can use [`drain`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain)
```Rust
for (a, &b) in out.iter_mut().zip(v.iter().skip(1)) {
  *a = b;
}
```
using [skip](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip).